### PR TITLE
[RLlib; Offline RL] CQL: Support multi-GPU/CPU setup and different learning rates for actor, critic, and alpha.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

This PR introduces - similar to `SAC` - multiple learning rates for `CQL`, namely one learning rate for each of the three optimizers (i.e. for actor, critic, and the hyperparameter alpha). Furthermore, it moves all forward passes from the learner into the module and therewith enables multi-learner setups. 
While `SAC` had already all forward passes moved into `_forward_train`, CQL missed the ones that were used for the `CQL` loss. This PR provides a complete setup.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/cql/cql.py [hunk 1]
```diff
@@ -84,6 +85,12 @@ def __init__(self, algo_class=None):
         self.lagrangian_thresh = 5.0
         self.min_q_weight = 5.0
         self.lr = 3e-4
```
Comment (sven1977) [inline rllib/algorithms/cql/cql.py:87]: set this to None :) by default

File: rllib/algorithms/cql/cql.py [hunk 2]
```diff
@@ -234,6 +241,28 @@ def validate(self) -> None:
...
+        else:
+            raise ValueError(
+                f"The framework {self.framework_str} is not supported. " "Use `torch`."
+            )
+
+    @property
+    def _model_config_auto_includes(self):
+        return super()._model_config_auto_includes | {
+            "num_actions": self.num_actions,
+            "_deterministic_loss": self._deterministic_loss,
```
Comment (sven1977) [inline rllib/algorithms/cql/cql.py:263]: nit: Let's remove this deterministic loss thing. It's a relic from a long time ago (2020) when I was trying to debug SAC on torch vs our old SAC on tf. It serves no real purpose and just bloats the code.

File: rllib/algorithms/cql/torch/cql_torch_rl_module.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        # Call the super method.
+        fwd_out = super()._forward_train(batch)
+
+        # Make sure we perform a "straight-through gradient" pass here,
+        # ignoring the gradients of the q-net, however, still recording
+        # the gradients of the policy net (which was used to rsample the actions used
+        # here). This is different from doing `.detach()` or `with torch.no_grads()`,
+        # as these two methds would fully block all gradient recordings, including
+        # the needed policy ones.
+        all_params = list(self.pi_encoder.parameters()) + list(self.pi.parameters())
```
Comment (sven1977) [inline rllib/algorithms/cql/torch/cql_torch_rl_module.py:31]: Wait, why `pi`? We need to block the q-net gradients (same as in SAC), correct?

Comment (sven1977) [reply rllib/algorithms/cql/torch/cql_torch_rl_module.py:31]:
```
        all_params = (
            list(self.qf.parameters())
            + list(self.qf_encoder.parameters())
            + list(self.qf_twin.parameters())
            + list(self.qf_twin_encoder.parameters())
        )
        for param in all_params:
            param.requires_grad = False
        output["q_curr"] = self.compute_q_values(q_batch_curr)
        for param in all_params:
            param.requires_grad = True
```

Comment (sven1977) [reply rllib/algorithms/cql/torch/cql_torch_rl_module.py:31]: ^ from SAC

File: rllib/algorithms/sac/torch/sac_torch_rl_module.py [hunk 4]
```diff
@@ -141,12 +141,12 @@ def _forward_train(self, batch: Dict) -> Dict[str, Any]:
         # here). This is different from doing `.detach()` or `with torch.no_grads()`,
         # as these two methds would fully block all gradient recordings, including
         # the needed policy ones.
-        all_params = (
-            list(self.qf.parameters())
-            + list(self.qf_encoder.parameters())
-            + list(self.qf_twin.parameters())
-            + list(self.qf_twin_encoder.parameters())
-        )
+        all_params = list(self.qf.parameters()) + list(self.qf_encoder.parameters())
```
Comment (sven1977) [inline rllib/algorithms/sac/torch/sac_torch_rl_module.py:144]: great catch!!

File: rllib/algorithms/cql/cql.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/cql/cql.py:87]: Yes, exactly. We discussed this in the other PR concerning `SAC`.

File: rllib/algorithms/cql/cql.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/cql.py:263]: Great!! That saves us many lines of code!

File: rllib/algorithms/cql/torch/cql_torch_rl_module.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/algorithms/cql/torch/cql_torch_rl_module.py:31]: Well, I don't think so. That was my first impression of it. But using this kind of backward pass let's the CQL loss rise without limit. Which makes sense: The additional loss term is now merely an added constant which is different from time to time, but does not change in regard to its own amount. Removing in this forward pass the `gradient=False` does make the CQL loss term decrease by time - as it should. 

Note, the SAC loss is computed earlier and uses in the `super()._forward_train` exactly the logic you posted above.

**Review by sven1977 — APPROVED** [review]
LGTM! Thanks @simonsays1980 !!
